### DISCUSSION TITLE
Gtk.GestureClick.OnReleased doesn't seem to work

### DISCUSSION THREAD

**Discussion by PlatinumLucario** [gircore/gir.core#1015] [Q&A] [answered]
I've been currently working on upgrading the Bassoon Jukebox project from GTK3 to GTK4. And I've noticed an issue when replacing the ``Gtk.Scale.ButtonReleaseEvent`` event with a ``Gtk.GestureClick.OnReleased`` event.

It doesn't work at all. What's odd is that ``Gtk.GestureClick.OnPressed`` works, so does the other ones.

So as a result, I'm currently stuck with this issue:
[Screencast_20240131_113914-4.webm](https://github.com/gircore/gir.core/assets/1906189/8e962a9b-160d-4722-b9b7-47f16e88b7b7)

And when I try to replace this code:
```cs
private bool CheckPlayback()
        {
            // We need to have an audio file
            if (Audio == null)
                return false;

            if (Audio.IsPlaying)
            {
                // Turns the MovingPlaybackSlider bool to false when not moving the playback slider
                if (!PlaybackSliderClick.IsActive() && !PlaybackSliderDrag.IsActive())
                {
                    MovingPlaybackSlider = false;
                }
                else
                {
                    MovingPlaybackSlider = true;
                    var frameClock = PlaybackSlider.GetFrameClock();
                    var frameTime = frameClock.GetFrameTime();
                    var frameCounter = frameClock.GetFrameCounter();
                    var timings = frameClock.GetTimings(frameCounter);
                    var value = (float)PlaybackSlider.GetValue();
                    //Audio.Cursor = (float)PlaybackSlider.GetValue();
                }
                // Adjust the slider if playing (and not scrubbing)
                if (!MovingPlaybackSlider && !PlaybackSliderClick.IsActive() && !PlaybackSliderDrag.IsActive())
                {
                    double playbackValue = Map(Audio.Cursor, 0, Audio.Duration.TotalSeconds, PlaybackSlider.Adjustment.Lower, PlaybackSlider.Adjustment.Upper);
                    PlaybackSlider.Adjustment.Value = playbackValue;
                }

                return true;
            }

            // If not, don't do anything
            PlayButton.Active = false;      // Un-toggle play
            return false;
        }
```
With this:
```cs
private bool CheckPlayback()
        {
            // We need to have an audio file
            if (Audio == null)
                return false;

            if (Audio.IsPlaying)
            {
                // // Turns the MovingPlaybackSlider bool to false when not moving the playback slider
                // if (!PlaybackSliderClick.IsActive() && !PlaybackSliderDrag.IsActive())
                // {
                //     MovingPlaybackSlider = false;
                // }
                // else
                // {
                //     MovingPlaybackSlider = true;
                //     var frameClock = PlaybackSlider.GetFrameClock();
                //     var frameTime = frameClock.GetFrameTime();
                //     var frameCounter = frameClock.GetFrameCounter();
                //     var timings = frameClock.GetTimings(frameCounter);
                //     var value = (float)PlaybackSlider.GetValue();
                //     //Audio.Cursor = (float)PlaybackSlider.GetValue();
                // }
                // Adjust the slider if playing (and not scrubbing)
                if (!MovingPlaybackSlider)
                {
                    double playbackValue = Map(Audio.Cursor, 0, Audio.Duration.TotalSeconds, PlaybackSlider.Adjustment.Lower, PlaybackSlider.Adjustment.Upper);
                    PlaybackSlider.Adjustment.Value = playbackValue;
                }

                return true;
            }

            // If not, don't do anything
            PlayButton.Active = false;      // Un-toggle play
            return false;
        }
```

It will do this instead:
[Screencast_20240131_141258-1.webm](https://github.com/gircore/gir.core/assets/1906189/505b70e1-056b-4a11-8afd-a71a3722d4eb)

Is there anything I'm doing wrong? Is this a bug in Gir.Core? Shouldn't GestureClick.OnRelease event trigger when there's no mouse click occurring?

Is there any workarounds? Any help would be greatly appreciated!

Also, if anyone needs to test it out for themselves, here's the source code:
[Bassoon.zip](https://github.com/gircore/gir.core/files/14106782/Bassoon.zip)

**Comment by badcel** [comment]
According to the [docs](https://docs.gtk.org/gtk4/signal.GestureClick.released.html) it should work.

Can you provide a minimal sample application?

**Comment by PlatinumLucario** [comment]
Well, this was interesting. I've managed to fix the issue. It turns out that the issue was that the slider was expecting an ``unpaired_release`` signal, not a ``released`` signal. While clicking on the part of the Gtk.Scale that the slider isn't on calls for an ``end`` signal instead.

So I added these in:
```cs
            PlaybackSliderClick.OnStopped += OnPlaybackSliderMouseUp;
            PlaybackSliderClick.OnCancel += OnPlaybackSliderMouseUp;
            PlaybackSliderClick.OnReleased += OnPlaybackSliderMouseUp;
            PlaybackSliderClick.OnUnpairedRelease += OnPlaybackSliderMouseUp;
            PlaybackSliderClick.OnEnd += OnPlaybackSliderMouseUp;
```
And now the slider moves properly and can be adjusted and then will proceed as normal without staying in place.

> According to the [docs](https://docs.gtk.org/gtk4/signal.GestureClick.released.html) it should work.
> 
> Can you provide a minimal sample application?

I mean, I could've just provided the JukeboxGtk4 application on it's own, with it's dependencies intact. Since it is a very small application with just a JukeboxGtk4.csproj and Jukebox.cs files.

**Comment by ozbenh** [comment, marked as answer]
So I couldn't fit it properly using your technique, because "stopped" and "end" arrive early while it's still sliding... In my case I was trying to stop the value being updated in the background by some other mechanism while "sliding".

The fix ended up easy once I figured things out. I had to attach the GtkGestureClick to the *parent* of the slider (I made a GtkBox just aroung the slider for that purpose) and set the GtkGestureClick propagation phase to GTK_PHASE_CAPTURE.

In C this is:  gtk_event_controller_set_propagation_phase(GTK_EVENT_CONTROLLER (gesture), GTK_PHASE_CAPTURE);

**Reply by PlatinumLucario** [reply]
That's interesting too, I'll definitely give that one a go as well.
